[![kevlar build status][travisbadge]](https://travis-ci.org/dib-lab/kevlar)
[![PyPI version][pypibadge]](https://pypi.python.org/pypi/biokevlar)
[![Test coverage][codecovbadge]](https://codecov.io/github/dib-lab/kevlar)
[![kevlar documentation][rtdbadge]](http://kevlar.readthedocs.io/en/latest/?badge=latest)
[![Docker build status][dockerbadge]](https://quay.io/repository/dib-lab/kevlar)
[![MIT licensed][licensebadge]](https://github.com/dib-lab/kevlar/blob/master/LICENSE)

<!-- <img src="docs/_static/morpheus-kevlar.jpg" alt=" What if I told you we don't need alignments to find variants?" width="400px" /> -->

# kevlar

Daniel Standage, 2016-2019  
https://kevlar.readthedocs.io

Welcome to **kevlar**, software for predicting *de novo* genetic variants without mapping reads to a reference genome!
kevlar's *k*-mer abundance based method calls single nucleotide variants (SNVs), multinucleotide variants (MNVs), insertion/deletion variants (indels), and structural variants (SVs) simultaneously with a single simple model.
This software is free for use under the MIT license.

You can checkout or download kevlar at our <a href='https://github.com/dib-lab/kevlar
Link to documentation: https://kevlar.readthedocs.io
'> GitHub </a>

<h1>Download Kevlar for regular use</h1>

If youre just looking to test kevlar and not edit its code for your use,
follow the steps below and enjoy!

<h3>Unix like OS:</h3>

    $ pip3 install git+https://github.com/dib-lab/khmer.git

    $ pip3 install biokevlar

<h3>Windows OS:</h3>
  
     ¯\_(ツ)_/¯
<h1>How to run Kevlar</h1>

<h1>Setup development environment for Kevlar</h1>




[travisbadge]: https://img.shields.io/travis/dib-lab/kevlar.svg
[pypibadge]: https://img.shields.io/pypi/v/biokevlar.svg
[codecovbadge]: https://img.shields.io/codecov/c/github/dib-lab/kevlar.svg
[rtdbadge]: https://readthedocs.org/projects/kevlar/badge/?version=latest&maxAge=900
[dockerbadge]: https://quay.io/repository/dib-lab/kevlar/status
[licensebadge]: https://img.shields.io/badge/license-MIT-blue.svg
